include/pybind11/pybind11.h

```diff
@@ -126,7 +126,8 @@ class cpp_function : public function {
         struct capture { remove_reference_t<Func> f; };
 
         /* Store the function including any extra state it might have (e.g. a lambda capture object) */
-        auto rec = make_function_record();
+        auto unique_rec = make_function_record();
```

Comment:
Comment from #2746:

- @YannickJadoul (regarding `function_record::data`/`function_record::free_data`): This can still leak, I have just realized.

  It's very annoying that we cannot just have a `std::unique_ptr<detail::function_record, void (*)(detail::function_record *)>`, with destruct as deleter. The issue is that we cannot do this before all the `char *`s are `strdup`ed, and that 1) is only done in `initialize_generic`, and 2) the `repr` calls during `strdup` can also throw.

Reply:
Making `make_function_record()` return a `py::capsule` already wouldn't help, right?

Reply:
That would have the same issues as `std::unique_ptr<detail::function_record, void (*)(detail::function_record *)>`, I think? I.e., when do you know for sure that those strings have been `strdup`ed?

Reply:
I think I fixed it; trying to demonstrate these leaks in #2746's 8e8fc6a2cfd5e8e8ce92f4a07217435490e6649a